---
title: "GposExperiment"
author:
- name: Robert Shear
  affiliation: Dana-Farber Cancer Institute
vignette: |
  %\VignetteIndexEntry{GposExperiment}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
  BiocStyle::html_document:
    toc_float: true
Package: GposExperiment
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction

The NET-seq assay generates count data at single nucleotide resolution. `r BiocStyle::Biocpkg("GPosExperiment")` extends for scenarios where vectors of counts at nucleotide resolutions is required for identified genomic regions of interest. `RangedSummarizedExperiment`

# Constructing a `GPosExperiment` object

```{r setup, include=FALSE}
library(GPosExperiment)
library(rtracklayer)
```

```{r define_helper.functions}
# TODO Is there a way to do this in existing R or Bioconductor packages?

#' matrix_apply - apply a function to every element in a matrix
#'
#' The matrix x is unrolled, function f is applied to the elements one at a
#' time and then the matrix is re-rolled.
#' 
#' @param x a matrix
#' @param f the function that operates on the elements
#' @param ... additional parameters to function f
#'
#' @return A matrix the same shape as x with row and column names preserved

matrix_apply <- function(x, f, ...) {
  result <- sapply(x, f, ...)
  dim(result) <- dim(x)
  dimnames(result) <- dimnames(x)
  result
}
```

```{r creating.NETseqData.class}

data_root <- file.path(system.file(package = "GPosExperiment"), "extdata")
# TODO - Better sample data - with real masks
genelist = rtracklayer::import(file.path(data_root, "genelist.gff3.bgz"), genome = "sacCer3")
which <- genelist
names(which) <- which$ID
which <- which[c("YAL062W", "YAL060W", "YAL059W")]
bam_fnames <- list.files(file.path(data_root),
                         pattern = "*[.]bam$", full.names = TRUE)

nsd <- mapply(function(id, fn) {
  print(id)
  print(fn)
  NETseqDataFromBAM(id, 
                         fn,
                         which, seqinfo(which))
  },
             id = basename(bam_fnames),
             fn = bam_fnames)
```

```{r}

e <- GPosExperiment(nsd, rowRanges = which)
```

```{r}
s <- scores(e, apply_mask = TRUE, zero_fill = TRUE)
n_mask <- matrix_apply(s, function(u) sum(is.na(u$score)))

n_s <- matrix_apply(s, function(u) sum(u$score, na.rm = TRUE))
n_mask
n_s


```
