---
title: "GposExperiment"
author:
- name: Robert Shear
vignette: |
  %\VignetteIndexEntry{GposExperiment}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
  BiocStyle::html_document:
    toc_float: true
Package: GposExperiment
bibliography: references.bib
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction

The NET-seq assay [@Churchman2011] generates count data at single nucleotide resolution. 
The `GenomicRanges::GRanges` class is a convenient structure for analyzing and operating on NET-seq assay data. Furthermore, the `GPos` extension of `GRanges`, which efficiently operates on elements of width 1, is particularly suitable for NET-seq data for scenarios where single-nucleotide resolution is needed. 
The `r BiocStyle::Biocpkg("GPosExperiment")` class extends `RangedSummarizedExperiment` for the convenient use of such `GPos` data.

# The `NETseqData` Class



# Constructing a `GPosExperiment` object

```{r setup, include=FALSE}
library(GPosExperiment)
library(rtracklayer)
```

```{r creating.NETseqData.class}

data_root <- file.path(system.file(package = "GPosExperiment"), "extdata")
# TODO - Better sample data - with real masks
genelist = rtracklayer::import(file.path(data_root, "genelist.gff3.bgz"), genome = "sacCer3")
which <- genelist
names(which) <- which$ID
which <- which[c("YAL062W", "YAL060W", "YAL059W")]
bam_fnames <- list.files(file.path(data_root),
                         pattern = "*[.]bam$", full.names = TRUE)

nsd <- mapply(function(id, fn) {
  print(id)
  print(fn)
  NETseqDataFromBAM(id, 
                         fn,
                         which, seqinfo(which))
  },
             id = basename(bam_fnames),
             fn = bam_fnames)
```

```{r}

e <- GPosExperiment(nsd, rowRanges = which)
```

```{r}
s <- scores(e, apply_mask = TRUE, zero_fill = TRUE)
n_mask <- apply(s, 1:2, function(u) sum(is.na(u[[1]]$score)))

n_s <- apply(s, 1:2, function(u) sum(u[[1]]$score, na.rm = TRUE))
n_mask
n_s


```
