---
title: "HRseqExperiment Proposal V0.1"
author:
- name: Robert Shear
output:
  BiocStyle::pdf_document:
    toc_float: true
  BiocStyle::html_document:
    toc_float: true
Package: HRseqExperiment
---

| *This document is a draft proposal for an extension to the `SummarizedExperiment` package.*
| *All the R names in this draft document are preliminary. Suggestions for alternative names are welcome. Consequently, the names are subject to change without notice.*
| 
| *We have changed the name of the package from `GPosExperiment` to `HRseqExperiment`. In either case, the purpose of this package is to make it convenient to process data addressed by genomic coordinate at nucleotide resolution. The `GPos` class in the `GenomicRanges` package is specifically designed to make the processing of just this type of data efficient. But the name is only meaningful from the software developer's point of view. We coin the term "HRseq" for high resolution sequencing, to include assays that are High Resolution SEQence data.*


# Introduction

The `HRseqExperiment` package is an extension of `RangedSummarizedExperiment` that facilitates convenient management of experimental genomic data.

`RangedSummarizedExperiment` stores assay as matrices, which are easy to use when each each matrix element is scalar. But when each observation consists of a vector of values, the manipulation and visualization of the data can become cumbersome.

`HRseqExperiment` addresses this obsticle by providing convenient capabilities for storing, retrieving and manipulating with the same style and convenience as scalar assays.

While there are minimal syntactic extensions to `RangedSummarizedExperiment`, the class internals need to take account of performance and memory requirements for real world applications.

# The `HResExperiment` Package

## Usage

The essential mechanism of `HResExperiment` is the creation of an `HRseqAssay` object, which associates the genomically indexed observations with each sample and specifies processing options. These observations are passed to `HRseqAssay` either as a `GRangesList` or as a list of filenames of genomic files that correspond to the samples. In either case, the name for each list element must match a sample identifier (`colName` value).

The parent virtual class `SummarizedExperiemnt::Assays` consists of a `SimpleList` that points to `matrix` objects, each of which is an assay. In `HResExperiment`, the setter functions `assays<-` and `assay<-` will also accept an `HRseqAssay` object.

The `SummarizedExperiment` `assay` (`assay<-`) methods return (store) a matrix that conforms to the dimensions of the `SummarizedExperiment` object that contains the assays.

The (proposed) `HResExperiment` subclass of `RangedSummarizedExperiment`, the underlying data of an `assay` is the equivalent of a list of `GRanges` objects. Each `HResExperiment` column (sample) corresponds to one element of the list. The rows of the matrix are derived from the  `GRanges` objects stored in `RangedSummarizedExperiment@rowRanges`, each row preforms the equivalent of `findOverlaps` to map the `GRanges`-like data to a specific matrix element.

Each of the matrix elements can be returned as a vector (logical, integer, numeric, character, double, factor), or as a `GRanges`, `GPos`, or `Rle` object. When returning a vector, certain representation alternative can be selected. For example, if `strand.aware` is `TRUE` and the `strand` for the row's `GRanges` is negative, then the vector will be reversed. Gaps in the assay source `GRanges` equivalent can be interpreted as zeros or NA's.

Additionally, the `assay` method has an extension to allow for element-by-element computations to be performed prior to returning the matrix value.


### HRseqAssay Constructor

```         
HRseqAssay(x = GRangesList(), 
  seqFilenameList = NULL, 
  strand.aware = FALSE, 
  zero.fill = FALSE, 
  mcol.name = "score", 
  value.type = numeric)
```

#### x - `GrangesList`

Each `GRanges` list element is the assay readout for a specific sample. The name of each list element must correspond to the name of a sample. If `seqinfo` is present, it must match the value of seqinfo of any `HRseqAssay` to which it is assigned. If present, the `seqFilenameList` parameter must be null.

#### seqFilenameList

An alternative to the first constructor argument, for loading `GRanges` objects from files. A list of fully qualified filenames, each of which can be read by `rtracklayer` to produce a `GRanges` object. For strand-specific assays encoded in formats that do not report strand, such as bedGraph, the list element will contain a character vector of two filenames, each identified by their strand, viz.

```         
fnl <- list(Sample12 = 
  c('+'='/n/lab12/sample12-netseq-pos.bedgraph', 
    '-'='/n/lab12/sample12-netseq-neg.bedgraph'))

print(fnl[["Sample12"]]['+'])
                                      + 
"/n/lab12/sample12-netseq-pos.bedgraph" 
```

#### strand.aware

`logical`. See [strand.aware].

#### zero.fill

`logical`. See [zero.fill].

#### mcol.name

Where the GRanges objects have more than one `mcol`, the name of the `mcol` that holds the assay values.

#### value.type

The data type to be returned for each element when retrieving the value. Accepted values are logical, integer, numeric, double, character, factor, GRanges, Gpos, and Rle.

### The `Assay()<-` Extension

`assay(x, i) <- HRseqAssay(â€¦)`

There is no change to the syntax of the `assay` setter function. While the `HRSeqAssay` object is not a matrix-like object, when associated with an `HRExperiment` object, it is capable of generating a matrix-like object that conforms to the requirements of assay.

### assay() Extension

```         
assay(hse, i = 1, formula = ., return.type = NULL, strand.aware = NULL)
```

#### hse

A `HResExperiment` object.

#### i

The assay number or name.

#### formula

If present, a formula that operates on each matrix element seperately.

#### return type

Override to the default return type declared in `HRSeqAssay`.

#### strand.aware

Override to `HRSeqAssay`.

## Theory of Operation

TODO

### Strand-aware Processing

TODO

### Missing Value Processing

TODO: Zero filling, NA filling, masks

### Formula Syntax

TODO

# An Example

```         
#> print(genelist)
GRanges object with 5 ranges and 2 metadata columns:
          seqnames      ranges strand |          curie        gene
             <Rle>   <IRanges>  <Rle> |    <character> <character>
  YAL068C     chrI   1807-2169      - | SGD:S000002142        PAU8
  YAL067C     chrI   7235-9016      - | SGD:S000000062        SEO1
  YAL064W     chrI 21566-21850      + | SGD:S000000060        <NA>
  YAL063C     chrI 24000-27968      - | SGD:S000000059        FLO9
  YAL062W     chrI 31567-32940      + | SGD:S000000058        GDH3
  -------
  seqinfo: 17 sequences (1 circular) from sacCer3 genome

#> str(samples)
 chr [1:2] "SRR12840066" "SRR12840067"
#> str(grl, max.level = 1)
List of 2
 $ SRR12840066:Formal class 'GRanges' [package "GenomicRanges"] with 7 slots
 $ SRR12840067:Formal class 'GRanges' [package "GenomicRanges"] with 7 slots
 
#> se <- SummarizedExperiment(rowRanges = genelist, colData = sample_names,
+            assays = list(HRseqAssay(samples, strand.aware = TRUE)))
#> print(assay(se, "score"))
        SRR12840066  SRR12840067 
YAL068C numeric,363  numeric,363 
YAL067C numeric,1782 numeric,1782
YAL064W numeric,285  numeric,285 
YAL063C numeric,3969 numeric,3969
YAL062W numeric,1374 numeric,1374
#>
#> print(assay(s, "score", formula = ~sum(.) ))
        SRR12840066 SRR12840067
YAL068C          20           4
YAL067C          38          19
YAL064W          15           3
YAL063C         386         209
YAL062W         464         201
```

## Another Example

A methylation profiling by Bisulfite-Seq experiment has produced a bigwig file for each of several samples. We have an annotated set of genes in which we are interested in the form of a  `GRanges` object.

```{r}
# Here's the data we need to build the HRseqExperiment
#
# sample_data - a DataFrame containing data on each sample
# bisulfite_readout_files - list of filenames of the bigwig files
#     parallel to the sample_data
# genes_of_interest - A Granges object with one row for each gene of interest
# 
#
# Now construct the HRseqExperiment object
#
# 
hre <- HRseqExperiment(assays=list(
			HRseqAssay(seqFIlenameList=bisulfite_readout_files,
				strand.aware=FALSE,
				zero.fill=FALSE)),
			rowRanges=genes_of_interest,
			 colData=sample_data)
```

The value of `assayNames(hre)` will be `score`, the default name of both the assay and the source metadata column name.

To retrieve the first region of interest and the first sample by integer index:
```
print(`assay(hre, "score")[1,1][[[1]])
 [1]    NA    NA    NA    NA 0.462 0.510 0.939 0.180 0.767 0.496 ... 

```
